server/chat-plugins/datasearch.ts

```diff
@@ -1266,19 +1283,13 @@ function runDexsearch(target: string, cmd: string, canAll: boolean, message: str
 			}
 			if (matched) continue;
 
-			const format = Object.entries(Dex.data.Rulesets).find(([a, f]) => f.mod === usedMod);
-			const formatStr = format ? format[1].name : 'gen9ou';
-			const ruleTable = Dex.formats.getRuleTable(Dex.formats.get(formatStr));
-			const additionalRules = [];
-			if (nationalSearch && !ruleTable.has('standardnatdex')) additionalRules.push('standardnatdex');
-			if (nationalSearch && ruleTable.valueRules.has('minsourcegen')) additionalRules.push('!!minsourcegen=3');
-			const validator = TeamValidator.get(`${formatStr}${additionalRules.length ? `@@@${additionalRules.join(',')}` : ''}`);
-			const pokemonSource = validator.allSources();
-			for (const move of Object.keys(alts.moves).map(x => mod.moves.get(x))) {
-				if (move.gen <= mod.gen && !validator.checkCanLearn(move, dex[mon], pokemonSource) === alts.moves[move.id]) {
+			for (const move of altsMoves) {
+				/** @ts-expect-error validator and pokemonSource won't be undefined if there's at least one move */
+				if (!validator.checkCanLearn(move, dex[mon], pokemonSource) === alts.moves[move.id]) {
```

Comment:
```suggestion
				if (!validator.checkCanLearn(move, dex[mon], pokemonSource) === alts.moves[move.id]) {
```
```suggestion
				if (validator && !validator.checkCanLearn(move, dex[mon], pokemonSource) === alts.moves[move.id]) {
```
We prefer not using ts-expect-error if we can avoid it. Which we can here.

Reply:
That's understandable. Here, however, if it's not initialized and it was supposed to be, this would hide the logic error. So, my thinking was that if someone did mess up the validator initialization logic, this would fail early and loudly during tests.
But I guess that during a test, this would produce wrong values anyway, and it would eventually get caught.

Reply:
In that case you would want to do

`if (!validator!.checkCanLearn(...`

which is what I would recommend

Reply:
I completely agree that it's better to crash than to produce incorrect results.

Reply:
Or the other approach is to lazily initialise `validator` here if it hasn't already been initialised, something along these lines:
```typescript
let validator;
...
    if (!validator) {
        ...
        validator = ...
    }
    if (!validator.checkCanLearn ...
```